libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/NavigationQueueContainer.java

```diff
@@ -0,0 +1,213 @@
+
+    NavigationMetricsWrapper.feedbackEvent(feedbackSessionState, checkRouteProgress(routeProgress),
+      feedbackEvent.getSessionState().eventLocation(), feedbackEvent.getDescription(),
+      feedbackEvent.getFeedbackType(), "", feedbackEvent.getFeedbackId(),
+      mapboxNavigation.obtainVendorId(), locationEngineName);
+  }
+
+  void sendRerouteEvent(SessionState sessionState) {
+    sessionState = sessionState.toBuilder()
+      .afterRerouteLocations(Arrays.asList(
+        locationBuffer.toArray(new Location[locationBuffer.size()])))
+      .build();
+
+    NavigationMetricsWrapper.rerouteEvent(sessionState, checkRouteProgress(routeProgress),
+      sessionState.eventLocation(), locationEngineName);
+
+    for (SessionState session : queuedRerouteEvents) {
+      queuedRerouteEvents.set(queuedRerouteEvents.indexOf(session),
+        session.toBuilder().lastRerouteDate(
+          sessionState.rerouteDate()
+        ).build());
+    }
+
+    mapboxNavigation.setSessionState(mapboxNavigation.getSessionState().toBuilder().lastRerouteDate(
+      sessionState.rerouteDate()
+    ).build());
+  }
+
+  private FeedbackEvent findQueuedFeedbackEvent(String feedbackId) {
+    for (FeedbackEvent feedbackEvent : queuedFeedbackEvents) {
+      if (feedbackEvent.getFeedbackId().equals(feedbackId)) {
+        return feedbackEvent;
+      }
+    }
+    return null;
+  }
+
+  private void updateFeedbackQueue() {
+    Iterator<FeedbackEvent> iterator = queuedFeedbackEvents.listIterator();
+    while (iterator.hasNext()) {
+      FeedbackEvent feedbackEvent = iterator.next();
+      if (feedbackEvent.getSessionState().eventLocation() != null
+        && feedbackEvent.getSessionState().eventLocation().equals(locationBuffer.peekFirst())
+        || TimeUtils.dateDiff(feedbackEvent.getSessionState().rerouteDate(), new Date(), TimeUnit.SECONDS)
+        > TWENTY_SECOND_INTERVAL) {
+        sendFeedbackEvent(feedbackEvent);
+        iterator.remove();
+      }
+    }
+  }
+
+  private void updateRerouteQueue() {
+    Iterator<SessionState> iterator = queuedRerouteEvents.listIterator();
+    while (iterator.hasNext()) {
+      SessionState sessionState = iterator.next();
+      if (sessionState.eventLocation() != null
+        && sessionState.eventLocation().equals(locationBuffer.peekFirst())
+        || TimeUtils.dateDiff(sessionState.rerouteDate(), new Date(), TimeUnit.SECONDS)
+        > TWENTY_SECOND_INTERVAL) {
+        sendRerouteEvent(sessionState);
+        iterator.remove();
+      }
+    }
+  }
+
+  void setLocationEngineName(String locationEngineName) {
+    this.locationEngineName = locationEngineName;
+  }
+
+  void setRouteProgress(RouteProgress routeProgress) {
+    this.routeProgress = routeProgress;
+  }
+
+  private RouteProgress checkRouteProgress(RouteProgress routeProgress) {
+    if (routeProgress != null) {
+      return routeProgress;
+    }
+    return obtainNewRouteProgress();
+  }
+
+  private RouteProgress obtainNewRouteProgress() {
```

Comment:
It looks like this will still throw NPE in the metrics wrapper.  Anything that uses `directionsRoute()` or `currentLegProgress()` will still crash.

It may be cleaner to wrap `RouteProgress` in a class (`MetricRouteProgress` for example) that only has getters for the information needed in `NavigationMetricsWrapper`.  So, if null, the class will still provide `0` or `""` to the wrapper class for the event depending on the variable needed.  

The same would apply to a null Location.  So a class that wraps `Location` and provides `0.0` `0.0` if the location it was constructed with is null.